app/rules/BucketRuleResource.scala

```diff
@@ -0,0 +1,80 @@
+package rules
+
+import play.api.libs.json.Json
+import com.amazonaws.services.s3.AmazonS3
+import com.amazonaws.services.s3.model.{ObjectMetadata, PutObjectRequest, S3Object}
+import java.io.InputStream
+import java.util.Date
+
+import com.amazonaws.{AmazonServiceException, SdkClientException}
+import model.RegexRule
+import utils.Loggable
+
+class BucketRuleResource (s3: AmazonS3, bucketName: String) extends Loggable {
+    private val RULES_KEY = "rules/typerighter-rules.json"
+
+    def serialiseAndUploadRules(rules: List[RegexRule]): Either[String, Date] = {
+        val ruleJson = Json.toJson(rules)
+        uploadRules(ruleJson.toString.getBytes(java.nio.charset.StandardCharsets.UTF_8.name))
+    }
```

Comment:
The structure here would be best as separate composable functions. A `serialiseRules` and a `uploadRules` which can then be nicely chained (taking advantage of the consistent left type) with something like:
```
for {
  ruleBlob <- serialise(rules)
  lastModified <- uploadRules(blob)
} yield lastModified
```

You might then put more into the comprehension so that it represents the whole flow of the program which I suspect would flatten out the nested case statement you have at the call site.